### PR TITLE
Laboratory work #2, Daria Moreva - 23FPL2

### PR DESCRIPTION
**Pull request by ddmoreva**
Text retrieval with BM25

### CODE REVIEW

File: seminars/practice_2_strings.py [hunk 1]
```diff
@@ -1,200 +0,0 @@
-"""
-Programming 2024
-Seminar 2
```
Comment (EIzharikov) [inline seminars/practice_2_strings.py:3]: revert this changes

File: requirements.txt [hunk 2]
```diff
@@ -1 +1,24 @@
...
+ast-comments==1.0.1
+black==22.6.0
+coverage[toml]==7.3.2
+doc8==1.1.1
+flake8-isort==6.0.0
+flake8-pyproject==1.2.3
+flake8==6.0.0
+ghapi==0.1.19
+mypy==1.1.1
+pydantic==1.10.7
```
Comment (EIzharikov) [inline requirements.txt:11]: revert this too :)

File: lab_2_retrieval_w_bm25/start.py [hunk 3]
```diff
@@ -1,29 +1,13 @@
...
 
 def main() -> None:
     """
     Launches an implementation
     """
-    paths_to_texts = [
-        "assets/fairytale_1.txt",
-        "assets/fairytale_2.txt",
-        "assets/fairytale_3.txt",
-        "assets/fairytale_4.txt",
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:15]: why? 😢

**Review by EIzharikov — COMMENTED** [review]
Nothing to review, but you created PR, thats good! Fix my comments and aim to 10! Good luck!

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -62,6 +95,25 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
...
+    if (not isinstance(vocab, list) or not isinstance(document_tokens, list) or
+            not all(isinstance(i, str) for i in document_tokens) or
+            not all(isinstance(k, str) for k in vocab)):
+        return None
+    dictionary_for_tf = {}
+    for word in vocab:
+        dictionary_for_tf[word] = 0.0
+        if word in document_tokens:
+            dictionary_for_tf[word] = document_tokens.count(word) / len(document_tokens)
+    for signal in document_tokens:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:111]: why signal

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -77,6 +129,24 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+            not all(isinstance(p, str) for p in vocab) or
+            not all(isinstance(k, list) and all(isinstance(elem, str)
+                                                for elem in k) for k in documents)):
+        return None
+    counter_for_documents = len(documents)
+    dictionary_for_idf = {}
+    for word in vocab:
+        dictionary_for_idf[word] = 0.0
+        counter = 0.0
+        for freq in documents:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:144]: why freq

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -77,6 +129,24 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+            if word in freq:
+                counter += 1.0
+            value = (counter_for_documents - counter + 0.5)/(counter + 0.5)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:147]: why + 0.5?

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -120,6 +206,31 @@ def calculate_bm25(
...
+            not isinstance(idf_document, dict)):
+        return None
+    if ((not all(isinstance(value, str) for value in vocab)) or
+            (not all(isinstance(value, str) for value in document)) or
+            (not all(isinstance(key, str) and isinstance(value, float) for key, value in idf_document.items()))):
+        return None
+    if ((not isinstance(k1, float)) or (not isinstance(b, float)) or
+            (not isinstance(avg_doc_len, float)) or (not isinstance(doc_len, int))):
+        return None
+    if (not 1.2 <= k1 <= 2.0) or (not 0 <= b <= 1):
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:221]: why 1.2 and 2.0?

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -120,6 +206,31 @@ def calculate_bm25(
...
+        return None
+    bm25 = {}
+    set_for_vocab = set(vocab)
+    set_for_doc = set(document)
+    all_need = list(set_for_doc.union(set_for_vocab))
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:226]: why all_need?

File: lab_2_retrieval_w_bm25/start.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 """
 Laboratory Work #2 starter
 """
-# pylint:disable=too-many-locals, unused-argument, unused-variable, too-many-branches, too-many-statements, duplicate-code
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:4]: why?

File: lab_2_retrieval_w_bm25/start.py [hunk 10]
```diff
@@ -30,5 +30,6 @@ def main() -> None:
     assert result, "Result is None"
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:30]: where is result?

**Review by EIzharikov — COMMENTED** [review]
Nice start! However, you should hurry up! Make CI green and fix all my comments. Good luck!

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -120,6 +207,34 @@ def calculate_bm25(
...
+            (not all(isinstance(key, str) and isinstance(value, float) for
+                     key, value in idf_document.items()))):
+        return None
+    if ((not isinstance(k1, float)) or (not isinstance(b, float)) or
+            ((not isinstance(avg_doc_len, float)) or (not isinstance(doc_len, int)) or
+            (isinstance(doc_len, bool)))):
+        return None
+    bm25 = {}
+
+    for word_in_doc in document:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:226]: can you do in dict init?

File: lab_2_retrieval_w_bm25/main.py [hunk 12]
```diff
@@ -120,6 +207,34 @@ def calculate_bm25(
...
+        bm25[word_in_doc] = 0.0
+
+    for word_in_idf in idf_document.keys():
+        if word_in_idf not in vocab:
+            return None
+        if word_in_idf not in bm25:
+            bm25[word_in_idf] = 0.0
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:233]: and after this you continue counting bm25? looks strange

File: lab_2_retrieval_w_bm25/main.py [hunk 13]
```diff
@@ -138,6 +253,35 @@ def rank_documents(
...
+
+    letters = tokenize(query)
+    if not isinstance(letters, list):
+        return None
+    without_stopwords = remove_stopwords(letters, stopwords)
+    if not isinstance(without_stopwords, list):
+        return None
+
+    list_with_index = []
+    for index_from_indexes, metrica in enumerate(indexes):
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:273]: what is ```metrica``` ?

File: lab_2_retrieval_w_bm25/main.py [hunk 14]
```diff
@@ -168,6 +312,33 @@ def calculate_bm25_with_cutoff(
 
     In case of corrupt input arguments, None is returned.
     """
+    # if not vocab or not document or not idf_document:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:315]: comments are redundant

**Review by EIzharikov — COMMENTED** [review]
Good progress. Make CI green and fix&**answer** all my comments. Good luck!

File: lab_2_retrieval_w_bm25/main.py [hunk 15]
```diff
@@ -62,6 +93,24 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
...
+    if not vocab or not document_tokens:
+        return None
+    if vocab is None or document_tokens is None:
+        return None
+    if (not isinstance(vocab, list) or not isinstance(document_tokens, list) or
+            not all(isinstance(i, str) for i in document_tokens) or
+            not all(isinstance(k, str) for k in vocab)):
+        return None
+    dictionary_for_tf = {}
+    for word in vocab:
```
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py:102]: ??

File: lab_2_retrieval_w_bm25/main.py [hunk 16]
```diff
@@ -62,6 +90,24 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
 
     In case of corrupt input arguments, None is returned.
     """
+    if not vocab or not document_tokens:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:93]: join checks

File: lab_2_retrieval_w_bm25/main.py [hunk 17]
```diff
@@ -92,6 +156,26 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
 
     In case of corrupt input arguments, None is returned.
     """
+    if not tf or not idf:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:159]: join checks

File: lab_2_retrieval_w_bm25/main.py [hunk 18]
```diff
@@ -192,6 +367,13 @@ def load_index(file_path: str) -> list[dict[str, float]] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(file_path, str) or not file_path:
+        return None
+    with open(file_path, 'r', encoding='utf-8') as file:
+        file_for_query: list[dict[str, float]] = json.load(file)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:373]: for what reason do you have this typing here?

File: lab_2_retrieval_w_bm25/main.py [hunk 19]
```diff
@@ -207,3 +389,16 @@ def calculate_spearman(rank: list[int], golden_rank: list[int]) -> float | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if (not rank or not golden_rank or not isinstance(rank, list)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:392]: join checks

File: lab_2_retrieval_w_bm25/start.py (hunk 9 above)
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/start.py:4]: ????

File: lab_2_retrieval_w_bm25/start.py [hunk 20]
```diff
@@ -24,9 +29,81 @@ def main() -> None:
...
+    tokenized_documents_without_stopwords = []
+    list_for_tf_idf = []
+    list_for_bm25 = []
+    list_for_bm25_without_cutoff = []
+    k1 = 1.5
+    b = 0.75
+    alpha = 0.2
+
+    for document in documents:
+        tokens: list[str] = tokenize(document) or []
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/start.py:45]: why typing?

**Review by EIzharikov — COMMENTED** [review]
You achieved 10, it is nice! Answer&fix **all** my comments and be ready for the next review. Good luck

**Review by EIzharikov — COMMENTED** [review]
I will not approve this lab until I get answers to all my questions and comments. Make CI green. Good luck

**Comment by ddmoreva** [comment]
i don`t understand, are all the comments I wrote not visible?...
